### ISSUE TITLE
Add Indexes/Sorts/Facet for Coordinate-Related Data

### ISSUE THREAD

**Issue by clarkepeterf** [project-lux/lux-marklogic#391]
**Problem Description:**
We would like to sort facets by the following:

- By Latitude
- By Longitude
- Has Map Coordinates

**Expected Behavior/Solution:**
Add sorting for those sorts + facet for Has Map Coordinates

**Requirements:**
I need to know what data to use for each.

As far as I can tell, location data can be either a POLYGON:

```
POLYGON ((-80.7338 40.03341, -80.73909 40.07567, -80.70894 40.10142, -80.70560999999999 40.15348, ...
```

or a POINT:

`POINT (-0.09305555555555554 51.51555555555556)`

I don't think we can sort by POLYGON, because we need to be able to grab a single latitude/ longitude from the data.

In order to sort, Marklogic would require separate fields for these, we may be able to use regex with an xpath, but I will have to check. Otherwise, we will need to have separate JSON properties to index. E.g.
```
"latitude": "-0.09305555555555554"
"longitude": "51.51555555555556"
```

For Has Map Coordinates, I think we should add a property similar to HasDigitalImage

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

- [ ] ~Wireframe/Mockup - Mike~
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [x] Backend requirements - TBD
- [x] Frontend requirements- TBD
- [ ] ~Are new regression tests required for QA - Amy~
- [x] Questions
- ~List of questions for discussions. Answers should be documented within the issue.~

**UAT/LUX Examples:**
- See FE example/TBD

**Dependencies/Blocks:**
- Blocking: https://github.com/project-lux/lux-frontend/issues/482

**Related Github Issues:**
- https://github.com/project-lux/lux-frontend/issues/396

**Related links:**
Options for calculating points within polygons:
- https://forum.weaviate.io/t/geocoordinates-sorting/763
- https://github.com/shapely/shapely <- fastest way to handle polygons
- https://rasterio.readthedocs.io/en/stable/

**Wireframe/Mockup:**
N/A

**Comment by roamye** [comment]
@clarkepeterf does this have a UAT or is it unreviewable? Also what else is missing for this to go into prioritization review? can it be moved to prioritization review?

**Comment by brent-hartwig** [comment]
Excerpt from an internal inquiry for sorting search results by geographical region:

> If some documents have points while others have regions, is there a norm for sorting these documents together?  I see we have https://docs.marklogic.com/geo.approxCenter, but that can't be worked into a sort specification and could be expensive in the context of Optic.  One or more points could be pre-calculated and indexed from a region, based on the customer's requirements (e.g., northern most point for longitude sort).  But just seeing what is typical.

I have not received a response yet.

**Comment by brent-hartwig** [comment]
@clarkepeterf, will more than this ticket be needed to unblock https://github.com/project-lux/lux-frontend/issues/482?  I'm wondering when a geospatial search pattern(s) will be needed.

cc: @prowns, @jffcamp

**Comment by roamye** [comment]
03/25 Team Mtg - 
- Licenses are required to use certain geospatial features needed to implement these indexes/sorts/facets. 
- Another approach can be used if we are unable to attain licenses, but it would prevent users from viewing overlaps between places. 

Next steps: 
- @jffcamp to review and confirm if we have these licenses. 
- defining requirements for this
- @clarkepeterf to do test indexes to understand how the data formats to geospatial indexing. 

_*this is a lower priority than MyCollections, but should still be moved to backlog when ready._

**Comment by brent-hartwig** [comment]
@clarkepeterf, is the licensing question around Yale's MarkLogic licenses?  Yale has the location services license.  From one of Blue's host status pages:

> conversion, failover, alerting, flexible replication, semantics, **location services**, English

**Comment by jffcamp** [comment]
We do NOT have the Advanced Geospatial license.

**Comment by brent-hartwig** [comment]
OK, so these are these restrictions you are referring to.

[Licensing Requirements for Geospatial Features](https://docs.progress.com/bundle/marklogic-server-use-search-11/page/topics/geospatial.html#id_34806):

> You must have an Advanced Geospatial License Option to use the following geospatial features:
> 
> * The functions geo:complex-polygon-contains, geo:complex-polygon-intersects, geo.complexPolygonContains, geo.complexPolygonIntersects.
> * Double precision coordinates, including wgs84/double, etrs89/double, and raw/double.
> * cts:reverse-query or cts.reverseQuery with geospatial constraints. This is sometimes called geo alerting.
>
> No other geospatial features or capabilities in MarkLogic require the Advanced Geospatial License Option.

**Comment by zachjschwartz** [comment]
@jffcamp @clarkepeterf Checking in to see if this issue is still in holding pattern. It looks like the ball was in Jeff's court when last updated in March.

**Comment by zachjschwartz** [comment]
@brent-hartwig @bshalke Do either of you know the status of this ticket?

**Comment by brent-hartwig** [comment]
@zachjschwartz, indexes alone?  No.  But we are tracking the need for a geo search pattern(s) which will surface index needs.
